spopt/locate/p_median.py

```diff
@@ -533,3 +535,517 @@ def solve(self, solver: pulp.LpSolver, results: bool = True):
+        """
+        Create the k nearest p-median problem from the sparse distance matrix.
+
+        Returns
+        -------
+
+        None
+        """
+        n_cli = self.sparse_matrix.shape[0]
+        r_cli = range(n_cli)
+        r_fac = range(self.sparse_matrix.shape[1])
+
+        self.weights = np.reshape(self.weights, (n_cli, 1))
+        # create and store the weighted sparse matrix
+        self.aij = self.sparse_matrix.multiply(self.weights).tocsr()
+
+        # create the model/problem
+        self.problem = pulp.LpProblem(self.name, pulp.LpMinimize)
+
+        # add all the 1)decision variable, 2)objective function, and 3)constraints
+
+        # Facility integer decision variable
+        FacilityModelBuilder.add_facility_integer_variable(self, r_fac, "y[{i}]")
+        fac_vars = getattr(self, "fac_vars")
+
+        # Placeholder facility decision variable
+        placeholder_vars = pulp.LpVariable.dicts(
+            "g", (i for i in r_cli), 0, 1, pulp.LpBinary
+        )
+        setattr(self, "placeholder_vars", placeholder_vars)
+
+        # Client assignment integer decision variables
+        row_indices, col_indices, values = find(self.aij)
+        cli_assgn_vars = pulp.LpVariable.dicts(
+            "z", [(i, j) for i, j in zip(row_indices, col_indices)], 0, 1, pulp.LpBinary
+        )
+        setattr(self, "cli_assgn_vars", cli_assgn_vars)
+
+        # Add the objective function
+        max_distance = self.aij.max(axis=1).toarray().flatten()
+        self.__add_obj(max_distance, r_cli, r_fac)
+
+        # Create the capacity constraints
+        if self.capacities is not None:
+            sorted_capacities = np.sort(self.capacities)
+            highest_possible_capacity = sorted_capacities[-self.p_facilities :].sum()
+            if highest_possible_capacity < self.ai_sum:
+                raise SpecificationError(
+                    "Problem is infeasible. The highest possible capacity "
+                    f"{highest_possible_capacity}, coming from the {self.p_facilities} "
+                    "sites with the highest capacity, is smaller than "
+                    f"the total demand {self.ai_sum}."
+                )
+            for j in col_indices:
+                self.problem += (
+                    pulp.lpSum(
+                        self.weights[i] * cli_assgn_vars.get((i, j), 0) for i in r_cli
+                    )
+                    <= fac_vars[j] * self.capacities[j]
+                )
+
+        # Create assignment constraints.
+        for i in r_cli:
+            self.problem += (
+                pulp.lpSum(cli_assgn_vars.get((i, j), 0) for j in set(col_indices))
+                + placeholder_vars[i]
+                == 1
+            )
+        # Create the facility constraint.
+        FacilityModelBuilder.add_facility_constraint(self, self.p_facilities)
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        k_list: np.array,
```

Comment:
Should the initial value for `k_list` default to an array of 1s? That seems the most logical case to initiate the  `KNearestPMedian`.

If so, perhaps we can set `k_list` as a keyword argument with the default value being `numpy.ones`. Then:

```python
if callable(k_list):
    k_list = k_list(gdf_demand.shape[0])
```

Reply:
I also think it's a great idea to set the default value for k_list to an array of one!

However, in Church's paper, he recommended to set the initial value of k as 5. I am not sure which way we should use now. Maybe we can discuss it in the meeting next week?